### PR TITLE
Reconcile node lbs on node deletion

### PR DESCRIPTION
**Pull request by maiqueb**
**- What this PR does and why is it needed**

services controller: reconcile on node deletion

A node load balancer is a load balancer associated to any of the following
services:
- service with NodePort set
- service with host-network endpoints
- service with ExternalTrafficPolicy=Local
- service with InternalTrafficPolicy=Local

This commit forces the reconciliation of the load balancers upon node
deletion. Since the node was deleted, the newly generated list of load
balancers will *not* feature the deleted node's load balancers, which will
lead to their deletion.

Leaking these load balancers causes an issue when the deleted node is re-added
(as part of a remediation procedure, for instance) since the newly created node
logical switch does not point to these node load balancers, thus breaking
connectivity of pods on the node to the services described in the list above.

closes: https://bugzilla.redhat.com/show_bug.cgi?id=2068910

**- Special notes for reviewers**

Forced service reconciliation when a node is deleted, to ensure the load balancers
associated to the node were deleted.

**- How to verify it**

1. Backup one of your node's configuration - `kubectl get node <nodeXYZ> -oyaml > nodexyz-backup.yaml`
2. Delete said node - `kubectl delete node <nodeXYZ>`
3. Ensure the node load balancers are **deleted** from the OVN NB database. These end with the `nodeXYZ` suffix
4. Add the node back - `kubectl apply -f nodexyz-backup.yaml`
5. Ensure the node load balancers are **found** from the OVN NB database. These end with the `nodeXYZ` suffix

**- Description for the changelog**

Force service reconciliation whenever a node is deleted.

### CODE REVIEW

**Comment by maiqueb** [comment]
Updated the last commit's msg in [this forced push](https://github.com/ovn-org/ovn-kubernetes/compare/44392acd0b8d432cdc0a9f5726615aa10c5fa5f3..09ac06b0d1b8631f427a1edfbfb6fa8d0bc1607d).

File: go-controller/pkg/testing/libovsdb/matchers.go [hunk 1]
```diff
@@ -203,41 +203,60 @@ func testDataEqual(x, y TestData, ignoreUUIDs bool) bool {
...
+func ContainData(expected ...TestData) gomegatypes.GomegaMatcher {
+	return haveData(false, true, expected, gomega.ContainElements)
+}
+
+// ContainDataIgnoringUUIDs succeeds if actual contains the passed in elements,
+// ignoring the elements UUIDs.
+// The ordering of the elements does not matter.
+func ContainDataIgnoringUUIDs(expected ...TestData) gomegatypes.GomegaMatcher {
+	return haveData(true, false, expected, gomega.ContainElements)
+}
```
Comment (jcaamano) [inline go-controller/pkg/testing/libovsdb/matchers.go:248]: We have been, or at least I have been, avoiding this type of matcher. It leads to not verifying the full contents of the database in a way that in-adverted changes can happen to the database that should have been verified not to happen. Even if you are sure that would not happen in your test, other people can start using this matcher without thinking things through.
Please, I won't be strict about it but reconsider your approach. What prevents in your test to verify the full contents of the database? Verifying the full contents of the database leads to a stronger test.

Comment (maiqueb) [reply go-controller/pkg/testing/libovsdb/matchers.go:248]: I understand your point. My motivation is to see a clear error when something goes wrong, instead of having to figure out what thing is missing/added in a huge haystack.

Would it make sense then to assert both ? First the "localized" matcher, which would give out the exact error, then the "general" matcher which would assert the full desired state of the DB ? It may seem redundant but I feel it does help out the developer.

Comment (jcaamano) [reply go-controller/pkg/testing/libovsdb/matchers.go:248]: > I understand your point. My motivation is to see a clear error when something goes wrong, instead of having to figure out what thing is missing/added in a huge haystack.

Certainly, processing the output of failed matcher over a big database requires a bit of practice. It still has a `missing things in expected` and `missing things in actual` that you could focus on? We might be able to customize its output to just that?

> Would it make sense then to assert both ? First the "localized" matcher, which would give out the exact error, then the "general" matcher which would assert the full desired state of the DB ? It may seem redundant but I feel it does help out the developer.

Would there be a way to express that with a single overall matcher that combines both things in a way that `ContainData` could not be used individually? 

Or perhaps it would be less harmful to have a `DoesNotContainData` matcher directly.

Comment (maiqueb) [reply go-controller/pkg/testing/libovsdb/matchers.go:248]: I feel the ginkgo way of doing this is not like that; that falls under the `Should` / `ShouldNot` / `To` / `NotTo` [assertion interface](https://github.com/ovn-org/ovn-kubernetes/blob/e19c20762c54f51615cd8ecff86eaea35e4b6efb/go-controller/vendor/github.com/onsi/gomega/gomega_dsl.go#L393).

Let me think this through a little bit, you gave me plenty to think about.

Comment (jcaamano) [reply go-controller/pkg/testing/libovsdb/matchers.go:248]: I don't wanna get in your way, just making quick suggestions to see if they make any sense to you.

Comment (maiqueb) [reply go-controller/pkg/testing/libovsdb/matchers.go:248]: I'm matching the original node balancer is not present **but** have also added matching against the final expected state of the cluster. 

If you think the focused matcher I've introduced is not required, I can remove the commit that has added it. 

If this approach seems preferable to you, I will also gladly squash the top-most commit.

Comment (maiqueb) [reply go-controller/pkg/testing/libovsdb/matchers.go:248]: @jcaamano ^

**Comment by jcaamano** [comment]
@maiqueb is the WIP commit WIP?

**Comment by maiqueb** [comment]
> @maiqueb is the WIP commit WIP?

Oh, no. That commit is solid - from our previous talk, it will surely be squashed into the previous one.

I just wanted to have the opportunity to present it separately. Sorry for confusing you. I kind of tried to give an heads-up in https://github.com/ovn-org/ovn-kubernetes/pull/3039#discussion_r901556820

**Review by jcaamano — CHANGES_REQUESTED** [review]

File: go-controller/pkg/ovn/controller/services/services_controller_test.go [hunk 2]
```diff
@@ -440,6 +518,13 @@ func TestSyncServices(t *testing.T) {
 			}
 
 			g.Eventually(controller.nbClient).Should(libovsdbtest.HaveData(tt.expectedDb))
+
+			if tt.nodeToDelete != nil {
+				controller.nodeTracker.removeNode(tt.nodeToDelete.name)
+				g.Expect(controller.syncService(namespacedServiceName(ns, serviceName))).To(gomega.Succeed())
+				g.Eventually(controller.nbClient).ShouldNot(libovsdbtest.ContainData(nodeRouterLoadBalancer(firstNode, nodePort, serviceName, ns, outport, nodeAEndpointIP, nodeBEndpointIP)))
+				g.Eventually(controller.nbClient).Should(libovsdbtest.HaveDataIgnoringUUIDs(tt.dbStateAfterDeleting))
```
[the remark this answers is not in the document]
Comment (maiqueb) [reply go-controller/pkg/ovn/controller/services/services_controller_test.go:526]: It doesn't work, with the following mismatch:

```
the missing elements were
            <[]*libovsdb.testDataMatcher | len:2, cap:2>: [
                {
                    expected: <*nbdb.LogicalSwitch | 0xc00149e1c0>{
                        UUID: "4d35a473-6481-4803-89a6-cb9026decb8a [switch-node-a]",
                        ACLs: nil,
                        Copp: nil,
                        DNSRecords: nil,
                        ExternalIDs: nil,
                        ForwardingGroups: nil,
                        LoadBalancer: [
                            "Service_testns/foo_TCP_cluster",
                        ],
                        LoadBalancerGroup: nil,
                        Name: "switch-node-a",
                        OtherConfig: nil,
                        Ports: nil,
                        QOSRules: nil,
                    },
                    ignoreUUID: false,
                },
                {
                    expected: <*nbdb.LogicalRouter | 0xc000146750>{
                        UUID: "700c91c2-0948-45fe-9aed-e8fa0f81227b [gr-node-a]",
                        Copp: nil,
                        Enabled: nil,
                        ExternalIDs: nil,
                        LoadBalancer: [
                            "Service_testns/foo_TCP_cluster",
                        ],
                        LoadBalancerGroup: nil,
                        Name: "gr-node-a",
                        Nat: nil,
                        Options: nil,
                        Policies: nil,
                        Ports: nil,
                        StaticRoutes: nil,
                    },
                    ignoreUUID: false,
                },
            ]
        the extra elements were
            <[]interface {} | len:2, cap:2>: [
                <*nbdb.LogicalRouter | 0xc000146b60>{
                    UUID: "700c91c2-0948-45fe-9aed-e8fa0f81227b [gr-node-a]",
                    Copp: nil,
                    Enabled: nil,
                    ExternalIDs: nil,
                    LoadBalancer: [
                        "eb32f513-9e84-49e3-8975-ad645116de1e",
                    ],
                    LoadBalancerGroup: nil,
                    Name: "gr-node-a",
                    Nat: nil,
                    Options: nil,
                    Policies: nil,
                    Ports: nil,
                    StaticRoutes: nil,
                },
                <*nbdb.LogicalSwitch | 0xc00149e620>{
                    UUID: "4d35a473-6481-4803-89a6-cb9026decb8a [switch-node-a]",
                    ACLs: nil,
                    Copp: nil,
                    DNSRecords: nil,
                    ExternalIDs: nil,
                    ForwardingGroups: nil,
                    LoadBalancer: [
                        "eb32f513-9e84-49e3-8975-ad645116de1e",
                    ],
                    LoadBalancerGroup: nil,
                    Name: "switch-node-a",
                    OtherConfig: nil,
                    Ports: nil,
                    QOSRules: nil,
                },
            ]
```

I.e. the associated load-balancers on the re-added node / router feature UUIDs, while we expect those to be names. 

If you suggest an alternative, I can try it out @jcaamano .

File: go-controller/pkg/ovn/controller/services/services_controller_test.go [hunk 3]
```diff
@@ -440,6 +518,13 @@ func TestSyncServices(t *testing.T) {
 			}
 
 			g.Eventually(controller.nbClient).Should(libovsdbtest.HaveData(tt.expectedDb))
+
+			if tt.nodeToDelete != nil {
+				controller.nodeTracker.removeNode(tt.nodeToDelete.name)
+				g.Expect(controller.syncService(namespacedServiceName(ns, serviceName))).To(gomega.Succeed())
+				g.Eventually(controller.nbClient).ShouldNot(libovsdbtest.ContainData(nodeRouterLoadBalancer(firstNode, nodePort, serviceName, ns, outport, nodeAEndpointIP, nodeBEndpointIP)))
```
[the remark this answers is not in the document]
Comment (maiqueb) [reply go-controller/pkg/ovn/controller/services/services_controller_test.go:525]: FIne, I'll remove it.

File: go-controller/pkg/ovn/controller/services/services_controller_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (maiqueb) [reply go-controller/pkg/ovn/controller/services/services_controller_test.go:526]: Somehow, seems the full re-sync put the UUIDs there. Does this make sense ?

[the remark this answers is not in the document]
Comment (maiqueb) [reply go-controller/pkg/ovn/controller/services/services_controller_test.go:526]: Expected LB (consistOf stanza - I get really confused about gingko sometimes ...):
```
{
    UUID: "c827591f-14d2-4b97-b015-9fdee6d81202 [Service_testns/foo_TCP_cluster]",
    ExternalIDs: {
        "k8s.ovn.org/kind": "Service",
        "k8s.ovn.org/owner": "testns/foo",
    },
    HealthCheck: nil,
    IPPortMappings: nil,
    Name: "Service_testns/foo_TCP_cluster",
    Options: {"skip_snat": "false", "event": "false", "reject": "true"},
    Protocol: "tcp",
    SelectionFields: nil,
    Vips: {
        "192.168.1.1:80": "10.128.0.2:3456,10.128.1.2:3456",
    },
},
```

Actual:
```
<*nbdb.LoadBalancer | 0xc00002b080>{
                    UUID: "c827591f-14d2-4b97-b015-9fdee6d81202 [Service_testns/foo_TCP_cluster]",
                    ExternalIDs: {
                        "k8s.ovn.org/owner": "testns/foo",
                        "k8s.ovn.org/kind": "Service",
                    },
                    HealthCheck: nil,
                    IPPortMappings: nil,
                    Name: "Service_testns/foo_TCP_cluster",
                    Options: {"reject": "true", "skip_snat": "false", "event": "false"},
                    Protocol: "tcp",
                    SelectionFields: nil,
                    Vips: {
                        "192.168.1.1:80": "10.128.0.2:3456,10.128.1.2:3456",
                    },
                },
```

The load balancers themselves **are the same** - otherwise they would be listed in the differences. 

The routers / switches pointing at them just feature names / UUIDs, which, as you say, is not the same. 

I don't know a way for this to work without using the `HaveDataIgnoringUUIDs` matcher. All ears though.

@jcaamano

[the remark this answers is not in the document]
Comment (jcaamano) [reply go-controller/pkg/ovn/controller/services/services_controller_test.go:526]: Something is wrong with the test or the code under test because
 `eb32f513-9e84-49e3-8975-ad645116de1e != c827591f-14d2-4b97-b015-9fdee6d81202`

Can you show the contents of the `eb32f513-9e84-49e3-8975-ad645116de1e`LB in actual.

[the remark this answers is not in the document]
Comment (maiqueb) [reply go-controller/pkg/ovn/controller/services/services_controller_test.go:526]: OK I see now what is happening, thanks for making me look deeper onto what is happening here.

Somehow, when the service is reconciled, the load balancers that are being removed from the deleted node logical switch / gateway router are the cluster-wide ones - `Service_testns/foo_TCP_cluster`. 

I somehow expected the load balancers that were dissociated from the logical switch / gw router to be the ones related to the node -  `Service_testns/foo_TCP_node_router+switch_node-a`. 
Thinking on this a bit more ... I think what really **should** happen is for **all** load balancers to be removed from the node's logical switch / gw router when the node is deleted.

I think this "doesn't matter" in the big picture, because the node's logical switches / gateway routers (which feature incorrect load balancers) will be deleted - as a result of the k8s node being deleted. Thus, having them feature the incorrect association does not matter much, since all pods are being evicted.

... never the less, this is a unit test. The service reconciler should work correctly no matter if the logical sw / gw router are deleted afterwards. I'll keep looking.

[the remark this answers is not in the document]
Comment (maiqueb) [reply go-controller/pkg/ovn/controller/services/services_controller_test.go:526]: Cleaning up the extra load balancer would require adding code to the production code to account for unit test behavior - the real OVSDB server would remove these LB entries from the logical switch / gw router (since they are weak references).

I do not think that is OK. 

I would rather push for allowing the `HaveDataIgnoringUUIDs` matcher, given the reasons explained above. 

Do you see an alternative @jcaamano ?

[the remark this answers is not in the document]
Comment (jcaamano) [reply go-controller/pkg/ovn/controller/services/services_controller_test.go:529]: I would get get the real UUIDs of the LBs that you expect to be removed from the DB before actually removing the node, and set them in the expected LS and LR at the end. I would still not use `HaveDataIgnoringUUIDs`.

File: go-controller/pkg/ovn/controller/services/services_controller_test.go [hunk 4]
```diff
@@ -628,3 +632,11 @@ func nodeConfig(nodeName string, nodeIP string) *nodeInfo {
...
 	}
 }
+
+func temporarilyEnableGomegaMaxLengthFormat() func() {
+	initialMaxLength := format.MaxLength
+	format.MaxLength = 0
+	return func() {
+		format.MaxLength = initialMaxLength
+	}
+}
```
[the remark this answers is not in the document]
Comment (maiqueb) [reply go-controller/pkg/ovn/controller/services/services_controller_test.go:642]: > One thing I had done in the past is having a `init()` function in `pkg/von/testing/libovsdb/matchers.go` to set this. It would be set every time the matcher was imported and that kinda made sense to me. But I guess not to everybody because it was removed :/

I didn't do that because I guess if it doesn't exist, people didn't want it. Which is basically the same you just said.

Thus, I've added it to the only place I needed it, hoping it wouldn't bother anyone else.

> 
> Anyway, the way you do it here is not very canonical for a test. Any reason you don't want to use the tear up/tear down mechanism of the testing framework. ginkgo has BeforeEach, etc....

I agree, but, whoever wrote this test, clearly didn't like ginkgo - it's not even imported, nor used.

Long story short, I don't want to re-write this test to work with ginkgo. I'd rather just drop the func.

[the remark this answers is not in the document]
Comment (jcaamano) [reply go-controller/pkg/ovn/controller/services/services_controller_test.go:642]: Ah, ginkgo was just an example. I checked the testing package that is used here and it has a cleanup function, but nothing very sophisticated. 

https://pkg.go.dev/testing#F.Cleanup

This is a bit of a nit but having something more similar to 
```
func setupTest(tb testing.TB) func(tb testing.TB) {...}
```
would probably look more familiar.

[the remark this answers is not in the document]
Comment (maiqueb) [reply go-controller/pkg/ovn/controller/services/services_controller_test.go:642]: I feel that belongs on a test framework.

I am simply trying to create a button that enables the full output of the diff to be printed on error - without pushing that to everyone else.

I will happily comply w/ using the `t.Cleanup` func though.

File: go-controller/pkg/ovn/controller/services/services_controller_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (maiqueb) [reply go-controller/pkg/ovn/controller/services/services_controller_test.go]: Would you take another look @jcaamano ? 

I think this covers it.

This "hack" could be removed if we somehow add weak reference cleanup to the test framework.

I could try to investigate that, and eventually (... hopefully) propose a follow-up PR for that.

File: go-controller/pkg/ovn/controller/services/services_controller_test.go [hunk 5]
```diff
@@ -469,133 +343,160 @@ func TestSyncServices(t *testing.T) {
...
+				nodeLogicalSwitch(nodeA,
+					loadBalancerClusterWideTCPServiceName(ns, serviceName)),
+				nodeLogicalSwitch(nodeB,
+					loadBalancerClusterWideTCPServiceName(ns, serviceName),
+					nodeSwitchRouterLoadBalancerName(nodeB, ns, serviceName)),
+				nodeLogicalRouter(nodeA,
+					loadBalancerClusterWideTCPServiceName(ns, serviceName)),
+				nodeLogicalRouter(nodeB,
+					loadBalancerClusterWideTCPServiceName(ns, serviceName),
+					nodeSwitchRouterLoadBalancerName(nodeB, ns, serviceName)),
```
Comment (jcaamano) [inline go-controller/pkg/ovn/controller/services/services_controller_test.go:499]: We shouldn't expect the cluster wide load balancers to be associated with the nodeA LR and LS, right? I know that you are patching them out below but it would be better to not have them here as well?

File: go-controller/pkg/ovn/controller/services/services_controller_test.go [hunk 6]
```diff
@@ -469,133 +343,160 @@ func TestSyncServices(t *testing.T) {
...
+				nodeLogicalRouter(nodeA,
+					loadBalancerClusterWideTCPServiceName(ns, serviceName),
+					nodeSwitchRouterLoadBalancerName(nodeA, ns, serviceName)),
+				nodeLogicalRouter(nodeB,
+					loadBalancerClusterWideTCPServiceName(ns, serviceName),
+					nodeSwitchRouterLoadBalancerName(nodeB, ns, serviceName)),
+			},
+			nodeToDelete: nodeConfig(nodeA, nodeAHostIP),
+			dbStateAfterDeleting: []libovsdbtest.TestData{&nbdb.LoadBalancer{
```
Comment (jcaamano) [inline go-controller/pkg/ovn/controller/services/services_controller_test.go:479]: LB in a new line here following the same format as the expected above?

File: go-controller/pkg/ovn/controller/services/services_controller_test.go [hunk 7]
```diff
@@ -627,7 +528,174 @@ func TestSyncServices(t *testing.T) {
...
+				// logical routers.
+				nodeLoadBalancerUUID, err := extractLoadBalancerRealUUID(
+					controller.nbClient,
+					nodeSwitchRouterLoadBalancerName(nodeA, ns, serviceName))
+				g.Expect(err).NotTo(gomega.HaveOccurred())
+				g.Expect(nodeLoadBalancerUUID).NotTo(gomega.BeEmpty())
+
+				g.Expect(controller.syncService(namespacedServiceName(ns, serviceName))).To(gomega.Succeed())
+
+				// here, we need to mutate the original expected data with the
```
Comment (jcaamano) [inline go-controller/pkg/ovn/controller/services/services_controller_test.go:548]: I don't think that `mutate` is the better word to use here, because it can be confused with the real ovsdb Mutate operation. `Patch` probably a better choice.

**Review by jcaamano — CHANGES_REQUESTED** [review]
Some more nits. Feel free to do the squashing you need.

File: go-controller/pkg/ovn/controller/services/services_controller_test.go (hunk 5 above)
Comment (maiqueb) [reply go-controller/pkg/ovn/controller/services/services_controller_test.go:499]: This is on nodeB (the one where nothing changes).

File: go-controller/pkg/ovn/controller/services/services_controller_test.go (hunk 6 above)
Comment (maiqueb) [reply go-controller/pkg/ovn/controller/services/services_controller_test.go]: Done.

File: go-controller/pkg/ovn/controller/services/services_controller_test.go (hunk 7 above)
Comment (maiqueb) [reply go-controller/pkg/ovn/controller/services/services_controller_test.go]: Done.

File: go-controller/pkg/ovn/controller/services/services_controller_test.go [hunk 8]
```diff
@@ -528,8 +532,30 @@ func TestSyncServices(t *testing.T) {
...
+				g.Expect(err).NotTo(gomega.HaveOccurred())
+				g.Expect(nodeLoadBalancerUUID).NotTo(gomega.BeEmpty())
+
 				g.Expect(controller.syncService(namespacedServiceName(ns, serviceName))).To(gomega.Succeed())
-				g.Eventually(controller.nbClient).Should(libovsdbtest.HaveDataIgnoringUUIDs(tt.dbStateAfterDeleting))
+
+				// here, we need to mutate the original expected data with the
+				// real load balancer UUID, on the logical switch / GW router
+				// of the deleted node, since the test library will be unable
+				// to correlate the name of the load balancer with its UUID.
```
[the remark this answers is not in the document]
Comment (maiqueb) [reply go-controller/pkg/ovn/controller/services/services_controller_test.go:550]: Done.

File: go-controller/pkg/ovn/controller/services/services_controller_test.go (hunk 5 above)
Comment (maiqueb) [reply go-controller/pkg/ovn/controller/services/services_controller_test.go]: Makes sense - at least *this* part of the test would be trivial to follow, and according to the expectations of the reader :) 

Good catch.

**Review by maiqueb — COMMENTED** [review]
Replying to @jcaamano 's review.

Thanks for the thorough review. Learnt quite a few things.

File: go-controller/pkg/ovn/controller/services/services_controller_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (maiqueb) [reply go-controller/pkg/ovn/controller/services/services_controller_test.go:550]: Is this better now ?

**Comment by jcaamano** [comment]
LGTM, thanks for the effort @maiqueb.

While we wait for the tests, I will give @martinkennelly a chance to take one extra final look.

**Comment by martinkennelly** [comment]
/lgtm
This was also educational for me. Thanks Jaime.
